test/common/filesystem/directory_test.cc

```diff
@@ -0,0 +1,228 @@
+  ScopedPathRemover s({file_path, dir_path});
+
+  TestUtility::createDirectory(dir_path);
+  { std::ofstream file(file_path); }
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"file", FileType::Regular},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, DirectoryWithOneDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string child_dir_path(dir_path + "/" + "sub_dir");
+  ScopedPathRemover s({child_dir_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  TestUtility::createDirectory(child_dir_path);
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"sub_dir", FileType::Directory},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, DirectoryWithFileInSubDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string child_dir_path(dir_path + "/" + "sub_dir");
+  const std::string child_file_path(child_dir_path + "/" + "file");
+  ScopedPathRemover s({child_file_path, child_dir_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  TestUtility::createDirectory(child_dir_path);
+  { std::ofstream file(child_file_path); }
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"sub_dir", FileType::Directory},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, false));
+}
+
+TEST(Directory, RecursionIntoSubDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string file_path(dir_path + "/" + "file");
+  const std::string child_dir_path(dir_path + "/" + "sub_dir");
+  const std::string child_file_path(child_dir_path + "/" + "sub_file");
+  ScopedPathRemover s({child_file_path, child_dir_path, file_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  TestUtility::createDirectory(child_dir_path);
+  { std::ofstream file(file_path); }
+  { std::ofstream file(child_file_path); }
+
+  EntrySet expected = {
+      {".", FileType::Directory},
+      {"..", FileType::Directory},
+      {"file", FileType::Regular},
+      {"sub_dir", FileType::Directory},
+      {"sub_dir/sub_file", FileType::Regular},
+      {"sub_dir/.", FileType::Directory},
+      {"sub_dir/..", FileType::Directory},
+  };
+  EXPECT_EQ(expected, getDirectoryContents(dir_path, true));
+}
+
+TEST(Directory, DirectoryWithFileAndDirectory) {
+  const std::string dir_path(TestEnvironment::temporaryPath("envoy_test"));
+  const std::string file_path(dir_path + "/" + "file");
+  const std::string child_dir_path(dir_path + "/" + "sub_dir");
+  ScopedPathRemover s({file_path, child_dir_path, dir_path});
+  TestUtility::createDirectory(dir_path);
+  TestUtility::createDirectory(child_dir_path);
+  { std::ofstream file(file_path); }
```

Comment:
There's a lot of repeated setup here, do you think we could move it to a test fixture and use `TEST_F`?